File: dungeons-1.21.11/src/main/java/fishmod/utils/Waypoint.java
package fishmod.utils;

import fishmod.utils.rendering.RenderUtils;
import net.minecraft.client.render.VertexConsumer;
import net.minecraft.client.util.math.MatrixStack;
import net.minecraft.util.math.Box;

public class Waypoint {
    private final double x, y, z, dx, dy, dz;
    private final float r, g, b, a;
    private final boolean throughWall;


    public Waypoint(double x, double y, double z, double dx, double dy, double dz, float r, float g, float b, float a, boolean throughWall) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.dx = dx;
        this.dy = dy;
        this.dz = dz;
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
        this.throughWall = throughWall;
    }

    public boolean samePosition(double x, double y, double z) {
        return this.x == x && this.y == y && this.z == z;
    }

    public boolean inRange(double x, double y, double z, double range) {
        double distanceSquared = (this.x - x) * (this.x - x) + (this.y - y) * (this.y - y) + (this.z - z) * (this.z - z);
        return distanceSquared < range * range;
    }

    public boolean isThroughWall() {
        return throughWall;
    }

    public void Render(VertexConsumer consumer, MatrixStack matrixStack) {
        RenderUtils.renderFilled(matrixStack, consumer, new Box(x, y, z, x + dx, y + dy, z + dz), new float[]{r, g, b, a});
    }

    @Override
    public String toString() {
        return "Waypoint{" +
                "x=" + x +
                ", y=" + y +
                ", z=" + z +
                ", dx=" + dx +
                ", dy=" + dy +
                ", dz=" + dz +
                ", r=" + r +
                ", g=" + g +
                ", b=" + b +
                ", a=" + a +
                ", throughWall=" + throughWall +
                '}';
    }
}

File: dungeons-1.21.11/src/main/kotlin/fishmod/utils/Waypoint.kt
package fishmod.utils

import fishmod.utils.rendering.RenderUtils
import net.minecraft.client.render.VertexConsumer
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.util.math.Box

class Waypoint(
    private val x: Double, private val y: Double, private val z: Double,
    private val dx: Double, private val dy: Double, private val dz: Double,
    private val r: Float, private val g: Float, private val b: Float, private val a: Float,
    private val throughWall: Boolean
) {

    fun samePosition(x: Double, y: Double, z: Double): Boolean {
        return this.x == x && this.y == y && this.z == z
    }

    fun inRange(x: Double, y: Double, z: Double, range: Double): Boolean {
        val distanceSquared = (this.x - x) * (this.x - x) + (this.y - y) * (this.y - y) + (this.z - z) * (this.z - z)
        return distanceSquared < range * range
    }

    fun isThroughWall(): Boolean {
        return throughWall
    }

    fun Render(consumer: VertexConsumer, matrixStack: MatrixStack) {
        RenderUtils.renderFilled(matrixStack, consumer, Box(x, y, z, x + dx, y + dy, z + dz), floatArrayOf(r, g, b, a))
    }

    override fun toString(): String {
        return "Waypoint{" +
                "x=" + x +
                ", y=" + y +
                ", z=" + z +
                ", dx=" + dx +
                ", dy=" + dy +
                ", dz=" + dz +
                ", r=" + r +
                ", g=" + g +
                ", b=" + b +
                ", a=" + a +
                ", throughWall=" + throughWall +
                '}'
    }
}
